core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory2.java

```diff
@@ -0,0 +1,262 @@
+    }
+
+    @Override
+    public void shutdown() {
+      for (Subchannel subchannel : getSubchannels()) {
+        subchannel.shutdown();
+      }
+    }
+
+    /**
+     * Updates picker with the list of active subchannels (state == READY).
+     */
+    private void updatePicker(@Nullable Status error) {
+      List<Subchannel> activeList = filterNonFailingSubchannels(getSubchannels());
+      helper.updatePicker(new Picker(activeList, error));
+    }
+
+    /**
+     * Filters out non-ready subchannels.
+     */
+    private static List<Subchannel> filterNonFailingSubchannels(
+        Collection<Subchannel> subchannels) {
+      List<Subchannel> readySubchannels = Lists.newArrayList();
+      for (Subchannel subchannel : subchannels) {
+        if (getSubchannelStateInfoRef(subchannel).get().getState() == READY) {
+          readySubchannels.add(subchannel);
+        }
+      }
+      return readySubchannels;
+    }
+
+    /**
+     * Converts list of {@link ResolvedServerInfoGroup} to {@link EquivalentAddressGroup} set.
+     */
+    private static Set<EquivalentAddressGroup> resolvedServerInfoGroupToEquivalentAddressGroup(
+        List<ResolvedServerInfoGroup> groupList) {
+      Set<EquivalentAddressGroup> addrs = Sets.newHashSet();
+      for (ResolvedServerInfoGroup group : groupList) {
+        for (ResolvedServerInfo server : group.getResolvedServerInfoList()) {
+          addrs.add(server.toEquivalentAddressGroup());
+        }
+      }
+      return addrs;
+    }
+
+    /**
+     * If all subchannels are TRANSIENT_FAILURE, return the Status associated with an arbitrary
+     * subchannel otherwise, return null.
+     */
+    @Nullable
+    private Status getAggregatedError() {
+      Status status = null;
+      for (Subchannel subchannel : getSubchannels()) {
+        ConnectivityStateInfo stateInfo = getSubchannelStateInfoRef(subchannel).get();
+        if (stateInfo.getState() != TRANSIENT_FAILURE) {
+          return null;
+        } else {
+          status = stateInfo.getStatus();
+        }
+      }
+      return status;
+    }
+
+    @VisibleForTesting
+    Collection<Subchannel> getSubchannels() {
+      return subchannels.values();
+    }
+
+    private static AtomicReference<ConnectivityStateInfo> getSubchannelStateInfoRef(
+        Subchannel subchannel) {
+      return checkNotNull(subchannel.getAttributes().get(STATE_INFO), "STATE_INFO");
+    }
+  }
+
+  @VisibleForTesting
+  static class Picker extends SubchannelPicker {
+    @GuardedBy("this")
+    final Iterator<Subchannel> subchannelIterator;
+    @Nullable
+    final Status status;
+
+    Picker(Iterable<Subchannel> iterable, @Nullable Status status) {
+      this.subchannelIterator = Iterables.cycle(iterable).iterator();
+      this.status = status;
+    }
+
+    @Override
+    public PickResult pickSubchannel(Attributes affinity, Metadata headers) {
+      // This will be true only if it's empty
+      if (subchannelIterator.hasNext()) {
```

Comment:
Ugh, yeah. Thanks. I've actually figured different solution to avoid calling `hasNext()` at all.